File: BEmu/HistoricalDataRequest/ElementHistoricFieldData.cs
//------------------------------------------------------------------------------
// <copyright project="BEmu" file="HistoricalDataRequest/ElementHistoricFieldData.cs" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

namespace BEmu.HistoricalDataRequest
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;

    internal class ElementHistoricFieldData : Element
    {
        private readonly Dictionary<string, Element> _fields; //make a dic

        public ElementHistoricFieldData(DateTime date, Dictionary<string, object> values)
        {
            this._fields = new Dictionary<string, Element>();

            Element elmDate = new ElementHistoricDateTime(date);
            this._fields.Add(elmDate.Name.ToString().ToUpper(), elmDate);

            foreach (var item in values)
            {
                if (item.Value is double)
                {
                    Element elmDouble = new ElementHistoricDouble(item.Key.ToUpper(), (double)item.Value);
                    this._fields.Add(elmDouble.Name.ToString(), elmDouble);
                }
            }
        }

        public override IEnumerable<Element> Elements { get { return this._fields.Values; } }

        public override Name Name { get { return new Name("fieldData"); } }
        public override int NumValues { get { return 1; } }
        public override int NumElements { get { return this._fields.Count; } }
        public override bool IsComplexType { get { return true; } }
        public override bool IsArray { get { return false; } }

        public override object this[int index] { get { return null; } }

        public override Element this[string name]
        {
            get
            {
                return this._fields[name.ToUpper()];
            }
        }

        public override object this[string name, int index]
        {
            get
            {
                if (index == 0)
                    return this[name].GetValue();
                else
                    return base[name, index];
            }
        }

        public override int GetElementAsInt32(string name)
        {
            return this[name].GetValueAsInt32();
        }

        public override Datetime GetElementAsDatetime(string name)
        {
            return this[name].GetValueAsDatetime();
        }

        public override Datetime GetElementAsDate(string name)
        {
            return this[name].GetValueAsDate();
        }

        public override Datetime GetElementAsTime(string name)
        {
            return this[name].GetValueAsTime();
        }

        public override string GetElementAsString(string name)
        {
            return this[name].GetValueAsString();
        }

        public override float GetElementAsFloat32(string name)
        {
            return this[name].GetValueAsFloat32();
        }

        public override double GetElementAsFloat64(string name)
        {
            return this[name].GetValueAsFloat64();
        }

        public override long GetElementAsInt64(string name)
        {
            return this[name].GetValueAsInt64();
        }

        public override bool HasElement(string name, bool excludeNullElements = false)
        {
            return this._fields.ContainsKey(name.ToUpper());
        }

        internal override StringBuilder PrettyPrint(int tabIndent)
        {
            string tabs = Types.IndentType.Indent(tabIndent);
            StringBuilder result = new StringBuilder();

            result.AppendFormat("{0}fieldData = {{{1}", tabs, Environment.NewLine);
            foreach (var item in this._fields.Values)
            {
                result.Append(item.PrettyPrint(tabIndent + 1));
            }
            result.AppendFormat("{0}}}{1}", tabs, Environment.NewLine);
            return result;
        }
    }
}

File: BEmuJava/bemu/src/main/java/com/bemu/BEmu/HistoricalDataRequest/ElementHistoricFieldData.java
//------------------------------------------------------------------------------
// <copyright project="BEmuJava" file="BEmu.HistoricalDataRequest.ElementHistoricFieldData.java" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

package com.bemu.BEmu.HistoricalDataRequest;

import com.bemu.BEmu.Datetime;
import com.bemu.BEmu.Name;

import java.util.Map;
import java.util.HashMap;

public class ElementHistoricFieldData extends ElementParent
{
	private final Map<String, ElementParent> _fields;
	
	ElementHistoricFieldData(Datetime date, Map<String, Object> values)
	{
		this._fields = new HashMap<String, ElementParent>();
		
		ElementParent elmDate = new ElementHistoricDateTime(date);
		this._fields.put(elmDate.name().toString().toUpperCase(), elmDate);
		
		for (Map.Entry<String, Object> item : values.entrySet())
		{
			if(item.getValue().getClass() == Double.class)
			{
				ElementParent elmDouble = new ElementHistoricDouble(item.getKey().toUpperCase(), (Double)item.getValue());
                this._fields.put(elmDouble.name().toString(), elmDouble);
			}
		}
	}
	
    public Name name()
    {
    	return new Name("fieldData");
    }
    
    public int numValues()
    {
    	return 1;
    }
    
    public int numElements()
    {
    	return this._fields.size();
    }
    
    public boolean isComplexType()
    {
    	return true;
    }
    
    public boolean isArray()
    {
    	return false;
    }
    
    public boolean isNull()
    {
    	return false;
    }
	
	public int getElementAsInt32(String name) throws Exception
	{
		return this._fields.get(name.toUpperCase()).getValueAsInt32();
	}
    
    public Datetime getElementAsDatetime(String name) throws Exception
    {
    	return this._fields.get(name.toUpperCase()).getValueAsDatetime();
    }
    
    public Datetime getElementAsDate(String name) throws Exception
    {
    	return this._fields.get(name.toUpperCase()).getValueAsDate();
    }
    
    public Datetime getElementAsTime(String name) throws Exception
    {
    	return this._fields.get(name.toUpperCase()).getValueAsTime();
    }
    
    public String getElementAsString(String name) throws Exception
    {
    	return this._fields.get(name.toUpperCase()).getValueAsString();
    }
	
	public float getElementAsFloat32(String name) throws Exception
	{
		return this._fields.get(name.toUpperCase()).getValueAsFloat32();
	}
	
	public double getElementAsFloat64(String name) throws Exception
	{
		return this._fields.get(name.toUpperCase()).getValueAsFloat64();
	}
	
	public long getElementAsInt64(String name) throws Exception
	{
		return this._fields.get(name.toUpperCase()).getValueAsInt64();
	}
	
	public boolean hasElement(String name)
	{
		return this._fields.containsKey(name);
	}
	
	public boolean hasElement(String name, boolean excludeNullElements)
	{
		return this._fields.containsKey(name);
	}
	
    protected StringBuilder prettyPrint(int tabIndent)
    {
        String tabs = com.bemu.BEmu.types.IndentType.Indent(tabIndent);
        StringBuilder result = new StringBuilder();

        result.append(String.format("%sfieldData = {%s", tabs, System.getProperty("line.separator")));
        
        for (Map.Entry<String, ElementParent> item : this._fields.entrySet())
        {
        	result.append(item.getValue().prettyPrint(tabIndent + 1));
        }

        result.append(String.format("%s}%s", tabs, System.getProperty("line.separator")));
        return result;
    }
	
}
